File: backend/app/infrastructure/repositories/knowledge_repository.py
"""
知识图谱仓储

提供知识节点和关系的 CRUD 操作
"""

from datetime import datetime

from sqlalchemy import and_, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.models.knowledge_node import NodeType, RelationType
from app.infrastructure.database.models import (
    KnowledgeNodeModel,
    KnowledgeRelationModel,
)


class KnowledgeRepository:
    """
    知识图谱仓储

    提供知识节点和关系的查询操作
    """

    def __init__(self, db: AsyncSession):
        self.db = db

    # =========================================================================
    # 知识节点查询
    # =========================================================================

    async def get_all_nodes(self) -> list[KnowledgeNodeModel]:
        """
        获取所有知识节点

        Returns:
            知识节点列表
        """
        stmt = select(KnowledgeNodeModel).order_by(KnowledgeNodeModel.created_at)
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def get_node_by_id(self, node_id: str) -> KnowledgeNodeModel | None:
        """
        根据 ID 获取知识节点

        Args:
            node_id: 节点 ID

        Returns:
            知识节点，不存在返回 None
        """
        stmt = select(KnowledgeNodeModel).where(KnowledgeNodeModel.id == node_id)
        result = await self.db.execute(stmt)
        return result.scalar_one_or_none()

    async def get_nodes_by_chapter(self, chapter: str) -> list[KnowledgeNodeModel]:
        """
        按章节筛选知识节点

        Args:
            chapter: 章节名称

        Returns:
            知识节点列表
        """
        stmt = (
            select(KnowledgeNodeModel)
            .where(KnowledgeNodeModel.chapter == chapter)
            .order_by(KnowledgeNodeModel.created_at)
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def get_nodes_by_type(self, node_type: NodeType) -> list[KnowledgeNodeModel]:
        """
        按类型筛选知识节点

        Args:
            node_type: 节点类型

        Returns:
            知识节点列表
        """
        stmt = (
            select(KnowledgeNodeModel)
            .where(KnowledgeNodeModel.node_type == node_type)
            .order_by(KnowledgeNodeModel.created_at)
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def search_nodes(self, keyword: str) -> list[KnowledgeNodeModel]:
        """
        按名称搜索知识节点

        Args:
            keyword: 搜索关键词

        Returns:
            知识节点列表
        """
        search_pattern = f"%{keyword}%"
        stmt = (
            select(KnowledgeNodeModel)
            .where(
                or_(
                    KnowledgeNodeModel.name.ilike(search_pattern),
                    KnowledgeNodeModel.name_en.ilike(search_pattern),
                    KnowledgeNodeModel.description.ilike(search_pattern),
                )
            )
            .order_by(KnowledgeNodeModel.created_at)
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def get_nodes_with_filters(
        self,
        *,
        chapter: str | None = None,
        node_type: NodeType | None = None,
        search: str | None = None,
    ) -> list[KnowledgeNodeModel]:
        """
        组合筛选知识节点

        Args:
            chapter: 章节名称（可选）
            node_type: 节点类型（可选）
            search: 搜索关键词（可选）

        Returns:
            知识节点列表
        """
        conditions = []

        if chapter:
            conditions.append(KnowledgeNodeModel.chapter == chapter)

        if node_type:
            conditions.append(KnowledgeNodeModel.node_type == node_type)

        if search:
            search_pattern = f"%{search}%"
            conditions.append(
                or_(
                    KnowledgeNodeModel.name.ilike(search_pattern),
                    KnowledgeNodeModel.name_en.ilike(search_pattern),
                    KnowledgeNodeModel.description.ilike(search_pattern),
                )
            )

        stmt = select(KnowledgeNodeModel)
        if conditions:
            stmt = stmt.where(and_(*conditions))

        stmt = stmt.order_by(KnowledgeNodeModel.created_at)

        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    # =========================================================================
    # 知识关系查询
    # =========================================================================

    async def get_all_relations(self) -> list[KnowledgeRelationModel]:
        """
        获取所有知识关系

        Returns:
            知识关系列表
        """
        stmt = select(KnowledgeRelationModel).order_by(
            KnowledgeRelationModel.created_at
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def get_relations_by_node(self, node_id: str) -> list[KnowledgeRelationModel]:
        """
        获取节点相关的所有关系（作为源节点或目标节点）

        Args:
            node_id: 节点 ID

        Returns:
            知识关系列表
        """
        stmt = select(KnowledgeRelationModel).where(
            or_(
                KnowledgeRelationModel.source_id == node_id,
                KnowledgeRelationModel.target_id == node_id,
            )
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def get_relations_by_type(
        self, relation_type: RelationType
    ) -> list[KnowledgeRelationModel]:
        """
        按关系类型筛选

        Args:
            relation_type: 关系类型

        Returns:
            知识关系列表
        """
        stmt = select(KnowledgeRelationModel).where(
            KnowledgeRelationModel.relation_type == relation_type
        )
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    # =========================================================================
    # 创建操作（用于数据初始化和管理）
    # =========================================================================

    async def create_node(self, node: KnowledgeNodeModel) -> KnowledgeNodeModel:
        """
        创建知识节点

        Args:
            node: 知识节点对象

        Returns:
            创建的节点
        """
        self.db.add(node)
        await self.db.flush()
        await self.db.refresh(node)
        return node

    async def create_relation(
        self, relation: KnowledgeRelationModel
    ) -> KnowledgeRelationModel:
        """
        创建知识关系

        Args:
            relation: 知识关系对象

        Returns:
            创建的关系
        """
        self.db.add(relation)
        await self.db.flush()
        await self.db.refresh(relation)
        return relation

    # =========================================================================
    # 更新和删除操作
    # =========================================================================

    async def update_node(
        self, node_id: str, **kwargs: object
    ) -> KnowledgeNodeModel | None:
        """
        更新知识节点

        Args:
            node_id: 节点 ID
            **kwargs: 要更新的字段

        Returns:
            更新后的节点，不存在返回 None
        """
        node = await self.get_node_by_id(node_id)
        if node is None:
            return None

        for key, value in kwargs.items():
            if hasattr(node, key):
                setattr(node, key, value)

        node.updated_at = datetime.now()
        await self.db.flush()
        await self.db.refresh(node)
        return node

    async def delete_node(self, node_id: str) -> bool:
        """
        删除知识节点（同时删除关联的关系）

        Args:
            node_id: 节点 ID

        Returns:
            是否成功删除
        """
        node = await self.get_node_by_id(node_id)
        if node is None:
            return False

        # 先删除关联的关系并 flush，确保外键约束不会阻止节点删除
        relations = await self.get_relations_by_node(node_id)
        if relations:
            for rel in relations:
                await self.db.delete(rel)
            await self.db.flush()

        await self.db.delete(node)
        await self.db.flush()
        return True

    async def get_relation_by_id(
        self, relation_id: str
    ) -> KnowledgeRelationModel | None:
        """根据 ID 获取知识关系"""
        stmt = select(KnowledgeRelationModel).where(
            KnowledgeRelationModel.id == relation_id
        )
        result = await self.db.execute(stmt)
        return result.scalar_one_or_none()

    async def update_relation(
        self, relation_id: str, **kwargs: object
    ) -> KnowledgeRelationModel | None:
        """
        更新知识关系

        Args:
            relation_id: 关系 ID
            **kwargs: 要更新的字段

        Returns:
            更新后的关系，不存在返回 None
        """
        relation = await self.get_relation_by_id(relation_id)
        if relation is None:
            return None

        for key, value in kwargs.items():
            if hasattr(relation, key):
                setattr(relation, key, value)

        await self.db.flush()
        await self.db.refresh(relation)
        return relation

    async def delete_relation(self, relation_id: str) -> bool:
        """
        删除知识关系

        Args:
            relation_id: 关系 ID

        Returns:
            是否成功删除
        """
        relation = await self.get_relation_by_id(relation_id)
        if relation is None:
            return False

        await self.db.delete(relation)
        await self.db.flush()
        return True

    # =========================================================================
    # 聚合查询
    # =========================================================================

    async def get_distinct_chapters(self) -> list[str]:
        """获取所有不重复的章节名称"""
        stmt = (
            select(KnowledgeNodeModel.chapter)
            .where(KnowledgeNodeModel.chapter.isnot(None))
            .where(KnowledgeNodeModel.chapter != "")
            .distinct()
            .order_by(KnowledgeNodeModel.chapter)
        )
        result = await self.db.execute(stmt)
        return [row[0] for row in result.all()]

    async def count_nodes(
        self,
        *,
        chapter: str | None = None,
        node_type: NodeType | None = None,
        search: str | None = None,
    ) -> int:
        """统计节点数量（支持筛选条件）"""
        conditions = self._build_filter_conditions(chapter, node_type, search)

        stmt = select(func.count(KnowledgeNodeModel.id))
        if conditions:
            stmt = stmt.where(and_(*conditions))

        result = await self.db.execute(stmt)
        return result.scalar() or 0

    async def get_nodes_paginated(
        self,
        *,
        chapter: str | None = None,
        node_type: NodeType | None = None,
        search: str | None = None,
        offset: int = 0,
        limit: int = 20,
    ) -> list[KnowledgeNodeModel]:
        """组合筛选知识节点（带分页）"""
        conditions = self._build_filter_conditions(chapter, node_type, search)

        stmt = select(KnowledgeNodeModel)
        if conditions:
            stmt = stmt.where(and_(*conditions))

        stmt = stmt.order_by(KnowledgeNodeModel.created_at).offset(offset).limit(limit)
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    # =========================================================================
    # 内部工具方法
    # =========================================================================

    @staticmethod
    def _build_filter_conditions(
        chapter: str | None,
        node_type: NodeType | None,
        search: str | None,
    ) -> list:
        """构建筛选条件列表（DRY: 复用于 count_nodes 和 get_nodes_paginated）"""
        conditions = []

        if chapter:
            conditions.append(KnowledgeNodeModel.chapter == chapter)
        if node_type:
            conditions.append(KnowledgeNodeModel.node_type == node_type)
        if search:
            search_pattern = f"%{search}%"
            conditions.append(
                or_(
                    KnowledgeNodeModel.name.ilike(search_pattern),
                    KnowledgeNodeModel.name_en.ilike(search_pattern),
                    KnowledgeNodeModel.description.ilike(search_pattern),
                )
            )

        return conditions

File: backend-go/internal/adapter/postgres/knowledge_repository.go
package postgres

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"

	knowledgeapp "mathstudy/backend-go/internal/application/knowledge"
)

const knowledgeNodeColumns = `
	id,
	name,
	name_en,
	node_type::text,
	description,
	chapter,
	section,
	difficulty,
	latex_formula,
	tags,
	created_at,
	updated_at`

const knowledgeRelationColumns = `
	kr.id,
	kr.source_id,
	kr.target_id,
	src.name,
	dst.name,
	kr.relation_type::text,
	kr.weight,
	kr.description,
	kr.created_at`

// KnowledgeRepository persists admin knowledge graph data in PostgreSQL.
type KnowledgeRepository struct {
	Repository
	beginner pgxTxBeginner
}

// NewKnowledgeRepository creates a PostgreSQL-backed knowledge repository.
func NewKnowledgeRepository(db Querier) (KnowledgeRepository, error) {
	base, err := NewRepository(db)
	if err != nil {
		return KnowledgeRepository{}, err
	}
	repo := KnowledgeRepository{Repository: base}
	if beginner, ok := db.(pgxTxBeginner); ok {
		repo.beginner = beginner
	}
	return repo, nil
}

// Stats returns aggregate knowledge graph counters.
func (r KnowledgeRepository) Stats(ctx context.Context) (knowledgeapp.Stats, error) {
	var stats knowledgeapp.Stats
	stats.TypeDistribution = map[string]int{}
	if err := r.DB().QueryRow(ctx, `SELECT count(id)::int FROM public.knowledge_nodes`).Scan(&stats.TotalNodes); err != nil {
		return knowledgeapp.Stats{}, err
	}
	if err := r.DB().QueryRow(ctx, `SELECT count(id)::int FROM public.knowledge_relations`).Scan(&stats.TotalRelations); err != nil {
		return knowledgeapp.Stats{}, err
	}
	if err := r.DB().QueryRow(ctx, `
		SELECT count(DISTINCT chapter)::int
		FROM public.knowledge_nodes
		WHERE chapter IS NOT NULL AND chapter <> ''`).Scan(&stats.ChaptersCount); err != nil {
		return knowledgeapp.Stats{}, err
	}
	rows, err := r.DB().Query(ctx, `
		SELECT node_type::text, count(id)::int
		FROM public.knowledge_nodes
		GROUP BY node_type
		ORDER BY node_type`)
	if err != nil {
		return knowledgeapp.Stats{}, err
	}
	defer rows.Close()
	for rows.Next() {
		var nodeType string
		var count int
		if err := rows.Scan(&nodeType, &count); err != nil {
			return knowledgeapp.Stats{}, err
		}
		stats.TypeDistribution[nodeTypeFromDB(nodeType)] = count
	}
	return stats, rows.Err()
}

// DistinctChapters returns sorted non-empty chapter names.
func (r KnowledgeRepository) DistinctChapters(ctx context.Context) ([]string, error) {
	rows, err := r.DB().Query(ctx, `
		SELECT DISTINCT chapter
		FROM public.knowledge_nodes
		WHERE chapter IS NOT NULL AND chapter <> ''
		ORDER BY chapter`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanStringColumn(rows)
}

// CountNodes counts nodes matching filters.
func (r KnowledgeRepository) CountNodes(ctx context.Context, filter knowledgeapp.NodeFilter) (int, error) {
	where, args := knowledgeNodeWhere(filter)
	var count int
	err := r.DB().QueryRow(ctx, `
		SELECT count(id)::int
		FROM public.knowledge_nodes
		WHERE `+where,
		args...,
	).Scan(&count)
	return count, err
}

// ListNodes returns a filtered node page.
func (r KnowledgeRepository) ListNodes(ctx context.Context, filter knowledgeapp.NodeFilter) ([]knowledgeapp.KnowledgeNode, error) {
	where, args := knowledgeNodeWhere(filter)
	offset := (filter.Page - 1) * filter.PageSize
	args = append(args, filter.PageSize, offset)
	limitPlaceholder := fmt.Sprintf("$%d", len(args)-1)
	offsetPlaceholder := fmt.Sprintf("$%d", len(args))
	rows, err := r.DB().Query(ctx, `
		SELECT `+knowledgeNodeColumns+`
		FROM public.knowledge_nodes
		WHERE `+where+`
		ORDER BY created_at
		LIMIT `+limitPlaceholder+` OFFSET `+offsetPlaceholder,
		args...,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	nodes := []knowledgeapp.KnowledgeNode{}
	for rows.Next() {
		node, err := scanAdminKnowledgeNode(rows)
		if err != nil {
			return nil, err
		}
		nodes = append(nodes, node)
	}
	return nodes, rows.Err()
}

// ListAllSimpleNodes returns compact node rows for relation selectors.
func (r KnowledgeRepository) ListAllSimpleNodes(ctx context.Context) ([]knowledgeapp.SimpleNode, error) {
	rows, err := r.DB().Query(ctx, `
		SELECT id, name, chapter, node_type::text
		FROM public.knowledge_nodes
		ORDER BY chapter NULLS LAST, name`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	nodes := []knowledgeapp.SimpleNode{}
	for rows.Next() {
		var node knowledgeapp.SimpleNode
		var chapter pgtype.Text
		var nodeType string
		if err := rows.Scan(&node.ID, &node.Name, &chapter, &nodeType); err != nil {
			return nil, err
		}
		if chapter.Valid {
			value := chapter.String
			node.Chapter = &value
		}
		value := nodeTypeFromDB(nodeType)
		node.NodeType = &value
		nodes = append(nodes, node)
	}
	return nodes, rows.Err()
}

// GetNode returns one knowledge node by ID.
func (r KnowledgeRepository) GetNode(ctx context.Context, nodeID string) (knowledgeapp.KnowledgeNode, bool, error) {
	row := r.DB().QueryRow(ctx, `
		SELECT `+knowledgeNodeColumns+`
		FROM public.knowledge_nodes
		WHERE id = $1`,
		nodeID,
	)
	node, err := scanAdminKnowledgeNode(row)
	if err != nil {
		if err == pgx.ErrNoRows {
			return knowledgeapp.KnowledgeNode{}, false, nil
		}
		return knowledgeapp.KnowledgeNode{}, false, err
	}
	return node, true, nil
}

// CreateNode inserts a knowledge node.
func (r KnowledgeRepository) CreateNode(ctx context.Context, input knowledgeapp.NodeInput, now time.Time) (knowledgeapp.KnowledgeNode, error) {
	nodeID, err := newUUID()
	if err != nil {
		return knowledgeapp.KnowledgeNode{}, err
	}
	tags, err := json.Marshal(input.Tags)
	if err != nil {
		return knowledgeapp.KnowledgeNode{}, err
	}
	row := r.DB().QueryRow(ctx, `
		INSERT INTO public.knowledge_nodes (
			id,
			name,
			name_en,
			node_type,
			description,
			chapter,
			section,
			difficulty,
			latex_formula,
			tags,
			created_at,
			updated_at
		)
		VALUES ($1, $2, $3, $4::public.nodetype, $5, $6, $7, $8, $9, $10::json, $11, $11)
		RETURNING `+knowledgeNodeColumns,
		nodeID,
		input.Name,
		input.NameEn,
		nodeTypeToDB(input.NodeType),
		input.Description,
		input.Chapter,
		input.Section,
		input.Difficulty,
		input.LatexFormula,
		string(tags),
		now,
	)
	return scanAdminKnowledgeNode(row)
}

// UpdateNode updates one knowledge node.
func (r KnowledgeRepository) UpdateNode(ctx context.Context, nodeID string, update knowledgeapp.NodeUpdate, now time.Time) (knowledgeapp.KnowledgeNode, bool, error) {
	sets := []string{}
	args := []any{nodeID}
	addSet := func(column string, value any) {
		args = append(args, value)
		sets = append(sets, fmt.Sprintf("%s = $%d", column, len(args)))
	}
	if update.Name != nil {
		addSet("name", *update.Name)
	}
	if update.NameEn != nil {
		addSet("name_en", *update.NameEn)
	}
	if update.NodeType != nil {
		addSet("node_type", nodeTypeToDB(*update.NodeType))
		sets[len(sets)-1] += "::public.nodetype"
	}
	if update.Description != nil {
		addSet("description", *update.Description)
	}
	if update.Chapter != nil {
		addSet("chapter", *update.Chapter)
	}
	if update.Section != nil {
		addSet("section", *update.Section)
	}
	if update.Difficulty != nil {
		addSet("difficulty", *update.Difficulty)
	}
	if update.LatexFormula != nil {
		addSet("latex_formula", *update.LatexFormula)
	}
	if update.Tags != nil {
		tags, err := json.Marshal(*update.Tags)
		if err != nil {
			return knowledgeapp.KnowledgeNode{}, false, err
		}
		addSet("tags", string(tags))
		sets[len(sets)-1] += "::json"
	}
	args = append(args, now)
	sets = append(sets, fmt.Sprintf("updated_at = $%d", len(args)))
	row := r.DB().QueryRow(ctx, `
		UPDATE public.knowledge_nodes
		SET `+strings.Join(sets, ", ")+`
		WHERE id = $1
		RETURNING `+knowledgeNodeColumns,
		args...,
	)
	node, err := scanAdminKnowledgeNode(row)
	if err != nil {
		if err == pgx.ErrNoRows {
			return knowledgeapp.KnowledgeNode{}, false, nil
		}
		return knowledgeapp.KnowledgeNode{}, false, err
	}
	return node, true, nil
}

// DeleteNode deletes a node and all adjacent relations.
func (r KnowledgeRepository) DeleteNode(ctx context.Context, nodeID string) (bool, error) {
	deleted := false
	err := r.withTx(ctx, func(tx KnowledgeRepository) error {
		if _, err := tx.DB().Exec(ctx, `
			DELETE FROM public.knowledge_relations
			WHERE source_id = $1 OR target_id = $1`,
			nodeID,
		); err != nil {
			return err
		}
		tag, err := tx.DB().Exec(ctx, `
			DELETE FROM public.knowledge_nodes
			WHERE id = $1`,
			nodeID,
		)
		if err != nil {
			return err
		}
		deleted = tag.RowsAffected() > 0
		return nil
	})
	return deleted, err
}

// NodeExists reports whether a node exists.
func (r KnowledgeRepository) NodeExists(ctx context.Context, nodeID string) (bool, error) {
	return r.Exists(ctx, `
		SELECT EXISTS(
			SELECT 1
			FROM public.knowledge_nodes
			WHERE id = $1
		)`,
		nodeID,
	)
}

// ListRelations returns relations with source and target node names.
func (r KnowledgeRepository) ListRelations(ctx context.Context, nodeID string) ([]knowledgeapp.KnowledgeRelation, error) {
	query := `
		SELECT ` + knowledgeRelationColumns + `
		FROM public.knowledge_relations kr
		LEFT JOIN public.knowledge_nodes src ON src.id = kr.source_id
		LEFT JOIN public.knowledge_nodes dst ON dst.id = kr.target_id`
	args := []any{}
	if strings.TrimSpace(nodeID) != "" {
		args = append(args, nodeID)
		query += `
		WHERE kr.source_id = $1 OR kr.target_id = $1`
	}
	query += `
		ORDER BY kr.created_at`
	rows, err := r.DB().Query(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	relations := []knowledgeapp.KnowledgeRelation{}
	for rows.Next() {
		relation, err := scanAdminKnowledgeRelation(rows)
		if err != nil {
			return nil, err
		}
		relations = append(relations, relation)
	}
	return relations, rows.Err()
}

// CreateRelation inserts a knowledge relation.
func (r KnowledgeRepository) CreateRelation(ctx context.Context, input knowledgeapp.RelationInput, now time.Time) (knowledgeapp.KnowledgeRelation, error) {
	relationID, err := newUUID()
	if err != nil {
		return knowledgeapp.KnowledgeRelation{}, err
	}
	if _, err := r.DB().Exec(ctx, `
		INSERT INTO public.knowledge_relations (
			id,
			source_id,
			target_id,
			relation_type,
			weight,
			description,
			created_at
		)
		VALUES ($1, $2, $3, $4::public.relationtype, $5, $6, $7)`,
		relationID,
		input.SourceID,
		input.TargetID,
		relationTypeToDB(input.RelationType),
		input.Weight,
		input.Description,
		now,
	); err != nil {
		return knowledgeapp.KnowledgeRelation{}, err
	}
	relation, ok, err := r.getRelationByID(ctx, relationID)
	if err != nil {
		return knowledgeapp.KnowledgeRelation{}, err
	}
	if !ok {
		return knowledgeapp.KnowledgeRelation{}, pgx.ErrNoRows
	}
	return relation, nil
}

// UpdateRelation updates one knowledge relation.
func (r KnowledgeRepository) UpdateRelation(ctx context.Context, relationID string, update knowledgeapp.RelationUpdate) (knowledgeapp.KnowledgeRelation, bool, error) {
	sets := []string{}
	args := []any{relationID}
	addSet := func(column string, value any) {
		args = append(args, value)
		sets = append(sets, fmt.Sprintf("%s = $%d", column, len(args)))
	}
	if update.RelationType != nil {
		addSet("relation_type", relationTypeToDB(*update.RelationType))
		sets[len(sets)-1] += "::public.relationtype"
	}
	if update.Weight != nil {
		addSet("weight", *update.Weight)
	}
	if update.Description != nil {
		addSet("description", *update.Description)
	}
	tag, err := r.DB().Exec(ctx, `
		UPDATE public.knowledge_relations
		SET `+strings.Join(sets, ", ")+`
		WHERE id = $1`,
		args...,
	)
	if err != nil {
		return knowledgeapp.KnowledgeRelation{}, false, err
	}
	if tag.RowsAffected() == 0 {
		return knowledgeapp.KnowledgeRelation{}, false, nil
	}
	relation, ok, err := r.getRelationByID(ctx, relationID)
	return relation, ok, err
}

// DeleteRelation deletes one relation.
func (r KnowledgeRepository) DeleteRelation(ctx context.Context, relationID string) (bool, error) {
	tag, err := r.DB().Exec(ctx, `
		DELETE FROM public.knowledge_relations
		WHERE id = $1`,
		relationID,
	)
	if err != nil {
		return false, err
	}
	return tag.RowsAffected() > 0, nil
}

func (r KnowledgeRepository) getRelationByID(ctx context.Context, relationID string) (knowledgeapp.KnowledgeRelation, bool, error) {
	row := r.DB().QueryRow(ctx, `
		SELECT `+knowledgeRelationColumns+`
		FROM public.knowledge_relations kr
		LEFT JOIN public.knowledge_nodes src ON src.id = kr.source_id
		LEFT JOIN public.knowledge_nodes dst ON dst.id = kr.target_id
		WHERE kr.id = $1`,
		relationID,
	)
	relation, err := scanAdminKnowledgeRelation(row)
	if err != nil {
		if err == pgx.ErrNoRows {
			return knowledgeapp.KnowledgeRelation{}, false, nil
		}
		return knowledgeapp.KnowledgeRelation{}, false, err
	}
	return relation, true, nil
}

func (r KnowledgeRepository) withTx(ctx context.Context, fn func(KnowledgeRepository) error) error {
	if fn == nil {
		return errors.New("knowledge transaction function is nil")
	}
	if r.beginner == nil {
		return fn(r)
	}
	tx, err := r.beginner.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return fmt.Errorf("begin knowledge transaction: %w", err)
	}
	base, err := NewRepository(tx)
	if err != nil {
		_ = tx.Rollback(ctx)
		return err
	}
	txRepo := KnowledgeRepository{Repository: base}
	if err := fn(txRepo); err != nil {
		if rollbackErr := tx.Rollback(ctx); rollbackErr != nil && !errors.Is(rollbackErr, pgx.ErrTxClosed) {
			return errors.Join(err, fmt.Errorf("rollback knowledge transaction: %w", rollbackErr))
		}
		return err
	}
	if err := tx.Commit(ctx); err != nil {
		if rollbackErr := tx.Rollback(ctx); rollbackErr != nil && !errors.Is(rollbackErr, pgx.ErrTxClosed) {
			return errors.Join(fmt.Errorf("commit knowledge transaction: %w", err), fmt.Errorf("rollback knowledge transaction: %w", rollbackErr))
		}
		return fmt.Errorf("commit knowledge transaction: %w", err)
	}
	return nil
}

func knowledgeNodeWhere(filter knowledgeapp.NodeFilter) (string, []any) {
	conditions := []string{"TRUE"}
	args := []any{}
	if strings.TrimSpace(filter.Chapter) != "" {
		args = append(args, filter.Chapter)
		conditions = append(conditions, fmt.Sprintf("chapter = $%d", len(args)))
	}
	if strings.TrimSpace(filter.NodeType) != "" {
		args = append(args, nodeTypeToDB(filter.NodeType))
		conditions = append(conditions, fmt.Sprintf("node_type = $%d::public.nodetype", len(args)))
	}
	if strings.TrimSpace(filter.Search) != "" {
		args = append(args, filter.Search)
		placeholder := fmt.Sprintf("$%d", len(args))
		conditions = append(conditions, "(name ILIKE '%' || "+placeholder+" || '%' OR name_en ILIKE '%' || "+placeholder+" || '%' OR description ILIKE '%' || "+placeholder+" || '%')")
	}
	return strings.Join(conditions, " AND "), args
}

func scanAdminKnowledgeNode(scanner rowScanner) (knowledgeapp.KnowledgeNode, error) {
	var node knowledgeapp.KnowledgeNode
	var nameEn pgtype.Text
	var chapter pgtype.Text
	var section pgtype.Text
	var latexFormula pgtype.Text
	var nodeType string
	var tagsRaw []byte
	if err := scanner.Scan(
		&node.ID,
		&node.Name,
		&nameEn,
		&nodeType,
		&node.Description,
		&chapter,
		&section,
		&node.Difficulty,
		&latexFormula,
		&tagsRaw,
		&node.CreatedAt,
		&node.UpdatedAt,
	); err != nil {
		return knowledgeapp.KnowledgeNode{}, err
	}
	node.NodeType = nodeTypeFromDB(nodeType)
	if nameEn.Valid {
		value := nameEn.String
		node.NameEn = &value
	}
	if chapter.Valid {
		value := chapter.String
		node.Chapter = &value
	}
	if section.Valid {
		value := section.String
		node.Section = &value
	}
	if latexFormula.Valid {
		value := latexFormula.String
		node.LatexFormula = &value
	}
	tags, err := decodeStringSlice(tagsRaw)
	if err != nil {
		return knowledgeapp.KnowledgeNode{}, fmt.Errorf("decode knowledge node tags: %w", err)
	}
	node.Tags = tags
	return node, nil
}

func scanAdminKnowledgeRelation(scanner rowScanner) (knowledgeapp.KnowledgeRelation, error) {
	var relation knowledgeapp.KnowledgeRelation
	var sourceName pgtype.Text
	var targetName pgtype.Text
	var relationType string
	var description pgtype.Text
	if err := scanner.Scan(
		&relation.ID,
		&relation.SourceID,
		&relation.TargetID,
		&sourceName,
		&targetName,
		&relationType,
		&relation.Weight,
		&description,
		&relation.CreatedAt,
	); err != nil {
		return knowledgeapp.KnowledgeRelation{}, err
	}
	if sourceName.Valid {
		value := sourceName.String
		relation.SourceName = &value
	}
	if targetName.Valid {
		value := targetName.String
		relation.TargetName = &value
	}
	relation.RelationType = relationTypeFromDB(relationType)
	if description.Valid {
		value := description.String
		relation.Description = &value
	}
	return relation, nil
}

func nodeTypeToDB(value string) string {
	return strings.ToUpper(strings.TrimSpace(value))
}

func nodeTypeFromDB(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}

func relationTypeToDB(value string) string {
	return strings.ToUpper(strings.TrimSpace(value))
}

func relationTypeFromDB(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}
